File: iosApp/iosApp/Pages/StopDetails/TripDetailsView.swift
//
//  TripDetailsView.swift
//  iosApp
//
//  Created by esimon on 11/14/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Combine
import Shared
import SwiftPhoenixClient
import SwiftUI

struct TripDetailsView: View {
    var tripFilter: TripDetailsPageFilter?

    var alertSummaries: [String: AlertSummary?]
    var context: TripDetailsViewModel.Context
    var now: EasternTimeInstant
    var routeAccents: TripRouteAccents

    let onOpenAlertDetails: (Shared.Alert) -> Void

    var errorBannerVM: IErrorBannerViewModel
    @ObservedObject var nearbyVM: NearbyViewModel
    var mapVM: IMapViewModel
    var tripDetailsVM: ITripDetailsViewModel

    @State var explainer: Explainer?
    @State var global: GlobalResponse?
    @State var tripDetailsVMState: TripDetailsViewModel.State?

    let analytics: Analytics
    var didLoadData: ((Self) -> Void)?
    let inspection = Inspection<Self>()

    init(
        tripFilter: TripDetailsPageFilter?,
        alertSummaries: [String: AlertSummary?],
        context: TripDetailsViewModel.Context,
        now: EasternTimeInstant,
        routeAccents: TripRouteAccents,
        onOpenAlertDetails: @escaping (Shared.Alert) -> Void,
        errorBannerVM: IErrorBannerViewModel,
        nearbyVM: NearbyViewModel,
        mapVM: IMapViewModel,
        tripDetailsVM: ITripDetailsViewModel = ViewModelDI().tripDetails,
        analytics: Analytics = AnalyticsProvider.shared
    ) {
        self.tripFilter = tripFilter
        self.alertSummaries = alertSummaries
        self.context = context
        self.now = now
        self.routeAccents = routeAccents
        self.onOpenAlertDetails = onOpenAlertDetails
        self.errorBannerVM = errorBannerVM
        self.nearbyVM = nearbyVM
        self.mapVM = mapVM
        self.tripDetailsVM = tripDetailsVM
        self.analytics = analytics
    }

    func getParentFor(_ stopId: String?) -> Stop? {
        if let global {
            global.getStop(stopId: stopId)?.resolveParent(global: global)
        } else { nil }
    }

    func onFollowTrip() {
        if let tripData = tripDetailsVMState?.tripData, let trip = tripData.trip, tripData.tripFilter == tripFilter {
            let dataFilter = tripData.tripFilter
            nearbyVM.pushNavEntry(.tripDetails(filter: TripDetailsPageFilter(
                tripId: dataFilter.tripId,
                vehicleId: dataFilter.vehicleId,
                routeId: trip.routeId,
                directionId: trip.directionId,
                stopId: dataFilter.stopId,
                stopSequence: dataFilter.stopSequence,
            )))
        }
    }

    var body: some View {
        content
            .explainer($explainer)
            .global($global, errorKey: "TripDetailsView")
            .onAppear {
                tripDetailsVMState = tripDetailsVM.models.value
                setVehicle(tripDetailsVM.selectedVehicleUpdates.value)
            }
            .task {
                for await models in tripDetailsVM.models {
                    tripDetailsVMState = models
                }
            }
            .task(id: global) {
                for await vehicleUpdate in tripDetailsVM.selectedVehicleUpdates {
                    setVehicle(vehicleUpdate)
                }
            }
            .onReceive(inspection.notice) { inspection.visit(self, $0) }
    }

    @ViewBuilder private var content: some View {
        VStack(spacing: 16) {
            let vehicle = tripDetailsVMState?.tripData?.vehicle
            if let tripFilter,
               tripFilter.vehicleId != nil ? vehicle != nil : true,
               let stops = tripDetailsVMState?.stopList,
               let tripData = tripDetailsVMState?.tripData,
               let trip = tripData.trip,
               tripData.tripFilter == tripFilter,
               tripData.tripPredictionsLoaded,
               let global,
               let route = global.getRoute(routeId: trip.routeId) {
                let terminalStop = getParentFor(trip.stopIds?.first)
                let vehicleStop = getParentFor(vehicle?.stopId)
                tripDetails(
                    tripFilter,
                    trip,
                    tripData.vehicleOnOtherTrip,
                    stops,
                    terminalStop,
                    vehicle,
                    vehicleStop,
                    route
                )
                .onAppear { didLoadData?(self) }
            } else if tripFilter != nil, let tripData = tripDetailsVMState?.tripData, tripData.vehicle != nil,
                      tripData.trip == nil {
                MissingTripCard(type: .notAvailable, routeAccents: routeAccents)
                    .padding(.horizontal, 14)
                    .padding(.top, 6)
            } else {
                loadingBody()
            }
        }
    }

    @ViewBuilder private func tripDetails(
        _ filter: TripDetailsPageFilter,
        _ trip: Trip,
        _ vehicleOnOtherTrip: Bool,
        _ stops: TripDetailsStopList,
        _ terminalStop: Stop?,
        _ vehicle: Vehicle?,
        _ vehicleStop: Stop?,
        _ route: Route,
    ) -> some View {
        let routeAccents = TripRouteAccents(route: route)
        let headerSpec: TripHeaderSpec? = {
            if let vehicle, let vehicleStop {
                let atTerminal = terminalStop != nil && terminalStop?.id == vehicleStop.id
                    && vehicle.currentStatus == .stoppedAt
                let entry = atTerminal ? stops.startTerminalEntry : stops.stops.first { entry in
                    entry.stop.id == vehicleStop.id
                }
                return vehicle.tripId == trip.id
                    ? .vehicle(vehicle, vehicleStop, entry, atTerminal)
                    : .finishingAnotherTrip
            } else if stops.stops.contains(where: { entry in entry.prediction != nil }) {
                return .noVehicle
            } else if let terminalStop, let terminalEntry = stops.startTerminalEntry {
                return .scheduled(terminalStop, terminalEntry)
            } else { return nil }
        }()

        let explainerType: ExplainerType? = switch headerSpec {
        case .scheduled: routeAccents.type != .ferry ? .noPrediction : nil
        case .finishingAnotherTrip: .finishingAnotherTrip
        case .noVehicle: .noVehicle
        default: nil
        }
        let onHeaderTap: (() -> Void)? = if let explainerType { {
            explainer = .init(type: explainerType, routeAccents: routeAccents)
        } } else { nil }

        // You can't open track this trip before the trip has started,
        // so if this is true, it means that the selected trip is complete
        if context == .tripDetails, vehicleOnOtherTrip {
            MissingTripCard(type: .complete, routeAccents: routeAccents).padding(.horizontal, 16)
        } else {
            VStack(spacing: 0) {
                tripHeaderCard(filter.stopId, trip, headerSpec, onHeaderTap, route, routeAccents).zIndex(1)
                    .padding(.horizontal, 6)
                    .padding(.top, 2)
                TripStops(
                    targetId: filter.stopId,
                    stops: stops,
                    stopSequence: filter.stopSequence?.intValue,
                    headerSpec: headerSpec,
                    now: now,
                    alertSummaries: alertSummaries,
                    onTapLink: onTapStop,
                    onOpenAlertDetails: onOpenAlertDetails,
                    route: route,
                    routeAccents: routeAccents,
                    global: global
                )
                .padding(.top, -56)
            }
        }
    }

    @ViewBuilder
    func tripHeaderCard(
        _ stopId: String,
        _ trip: Trip,
        _ spec: TripHeaderSpec?,
        _ onTap: (() -> Void)?,
        _ route: Route,
        _ routeAccents: TripRouteAccents,
    ) -> some View {
        if let spec {
            TripHeaderCard(
                spec: spec,
                trip: trip,
                targetId: stopId,
                route: route,
                routeAccents: routeAccents,
                onTap: onTap,
                now: now,
                onFollowTrip: context != .tripDetails ? onFollowTrip : nil,
            )
        }
    }

    @ViewBuilder private func loadingBody() -> some View {
        let placeholderInfo = LoadingPlaceholders.shared.tripDetailsInfo()
        tripDetails(
            .init(
                stopId: "",
                stopFilter: .init(routeId: Route.Id(""), directionId: 0),
                tripFilter: .init(tripId: "", vehicleId: nil, stopSequence: nil, selectionLock: false)
            ),
            placeholderInfo.trip,
            false,
            placeholderInfo.stops,
            nil,
            placeholderInfo.vehicle,
            placeholderInfo.vehicleStop,
            placeholderInfo.route,
        ).loadingPlaceholder()
    }

    private func setVehicle(_ vehicle: Vehicle?) {
        let stop = global?.getStop(stopId: nearbyVM.navigationStack.lastStopId)
        let stopFilter = nearbyVM.navigationStack.lastStopDetailsFilter
        let tripDetailsFilter = nearbyVM.navigationStack.lastTripDetailsFilter
        if let tripDetailsFilter {
            mapVM.selectedTrip(
                stopFilter: stopFilter,
                stop: stop,
                tripFilter: tripDetailsFilter,
                vehicle: vehicle,
                follow: false
            )
        }
    }

    func onTapStop(stop: TripDetailsStopList.Entry) {
        let parentStop = if let global { stop.stop.resolveParent(global: global) }
        else { stop.stop }
        nearbyVM.appendNavEntry(.stopDetails(stopId: parentStop.id, stopFilter: nil, tripFilter: nil))
        analytics.tappedDownstreamStop(
            routeId: tripDetailsVMState?.tripData?.trip?.routeId,
            stopId: stop.stop.id,
            tripId: tripFilter?.tripId ?? "",
            connectingRouteId: nil
        )
    }
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/stopDetails/TripDetailsView.kt
package com.mbta.tid.mbta_app.android.stopDetails

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.ui.zIndex
import androidx.lifecycle.compose.LifecycleResumeEffect
import com.mbta.tid.mbta_app.analytics.Analytics
import com.mbta.tid.mbta_app.android.ModalRoutes
import com.mbta.tid.mbta_app.android.state.getGlobalData
import com.mbta.tid.mbta_app.android.tripDetails.MissingTripCard
import com.mbta.tid.mbta_app.android.tripDetails.MissingTripCardType
import com.mbta.tid.mbta_app.android.util.IsLoadingSheetContents
import com.mbta.tid.mbta_app.android.util.modifiers.loadingShimmer
import com.mbta.tid.mbta_app.model.Alert
import com.mbta.tid.mbta_app.model.AlertSummary
import com.mbta.tid.mbta_app.model.LoadingPlaceholders
import com.mbta.tid.mbta_app.model.ObjectCollectionBuilder
import com.mbta.tid.mbta_app.model.Route
import com.mbta.tid.mbta_app.model.RouteType
import com.mbta.tid.mbta_app.model.Stop
import com.mbta.tid.mbta_app.model.Trip
import com.mbta.tid.mbta_app.model.TripDetailsPageFilter
import com.mbta.tid.mbta_app.model.TripDetailsStopList
import com.mbta.tid.mbta_app.model.response.AlertsStreamDataResponse
import com.mbta.tid.mbta_app.model.response.GlobalResponse
import com.mbta.tid.mbta_app.model.stopDetailsPage.ExplainerType
import com.mbta.tid.mbta_app.model.stopDetailsPage.TripData
import com.mbta.tid.mbta_app.model.stopDetailsPage.TripHeaderSpec
import com.mbta.tid.mbta_app.routes.SheetRoutes
import com.mbta.tid.mbta_app.utils.EasternTimeInstant
import com.mbta.tid.mbta_app.viewModel.ITripDetailsViewModel
import com.mbta.tid.mbta_app.viewModel.TripDetailsViewModel
import io.sentry.kotlin.multiplatform.Sentry
import org.koin.compose.koinInject

@Composable
fun TripDetailsView(
    tripFilter: TripDetailsPageFilter?,
    allAlerts: AlertsStreamDataResponse?,
    alertSummaries: Map<String, AlertSummary?>,
    onOpenAlertDetails: (Alert) -> Unit,
    openSheetRoute: (SheetRoutes) -> Unit,
    openModal: (ModalRoutes) -> Unit,
    now: EasternTimeInstant,
    routeAccents: TripRouteAccents,
    isTripDetailsPage: Boolean,
    modifier: Modifier = Modifier,
    tripDetailsVM: ITripDetailsViewModel = koinInject(),
    analytics: Analytics = koinInject(),
) {
    val globalResponse: GlobalResponse? = getGlobalData(errorKey = "TripDetailsView")
    val state by tripDetailsVM.models.collectAsState()
    val tripData: TripData? = state.tripData
    val stopList = state.stopList
    val vehicle = tripData?.vehicle

    val tripDetailsContext: TripDetailsViewModel.Context =
        if (isTripDetailsPage) TripDetailsViewModel.Context.TripDetails
        else TripDetailsViewModel.Context.StopDetails

    LaunchedEffect(tripFilter) { tripDetailsVM.setFilters(tripFilter) }
    LaunchedEffect(allAlerts) { tripDetailsVM.setAlerts(allAlerts) }

    LaunchedEffect(Unit) {
        tripDetailsVM.setContext(tripDetailsContext)
        tripDetailsVM.setActive(active = true, wasSentToBackground = false)
    }

    LifecycleResumeEffect(Unit) {
        tripDetailsVM.setActive(active = true, wasSentToBackground = false)
        onPauseOrDispose { tripDetailsVM.setActive(active = false, wasSentToBackground = true) }
    }

    fun getParentFor(stopId: String?, globalResponse: GlobalResponse): Stop? {
        return stopId.let { globalResponse.getStop(stopId)?.resolveParent(globalResponse) }
    }

    fun onTapStop(stop: TripDetailsStopList.Entry) {
        val parentStationId =
            globalResponse?.getStop(stop.stop.id)?.resolveParent(globalResponse)?.id ?: stop.stop.id
        openSheetRoute(SheetRoutes.StopDetails(parentStationId, null, null))
        analytics.tappedDownstreamStop(
            routeId = tripData?.trip?.routeId,
            stopId = parentStationId,
            tripId = tripFilter?.tripId ?: "",
            connectingRouteId = null,
        )
    }

    val trip = tripData?.trip
    if (
        tripFilter != null &&
            tripData != null &&
            trip != null &&
            globalResponse != null &&
            stopList != null &&
            tripData.tripFilter == tripFilter &&
            stopList.trip.id == trip.id
    ) {
        val route =
            globalResponse.getRoute(trip.routeId)
                ?: run {
                    Sentry.captureMessage("Trip ${trip.id} on unknown route ${trip.routeId}")
                    return
                }
        val terminalStop = getParentFor(trip.stopIds?.firstOrNull(), globalResponse)
        val vehicleStop = vehicle?.let { getParentFor(it.stopId, globalResponse) }
        val tripId = tripFilter.tripId
        val headerSpec: TripHeaderSpec? =
            TripHeaderSpec.getSpec(tripId, stopList, terminalStop, vehicle, vehicleStop)

        val explainerType: ExplainerType? =
            when (headerSpec) {
                is TripHeaderSpec.Scheduled ->
                    if (route.type != RouteType.FERRY) {
                        ExplainerType.NoPrediction
                    } else {
                        null
                    }
                is TripHeaderSpec.FinishingAnotherTrip -> ExplainerType.FinishingAnotherTrip
                is TripHeaderSpec.NoVehicle -> ExplainerType.NoVehicle
                else -> null
            }
        val onHeaderTap: (() -> Unit)? =
            if (explainerType != null) {
                { openModal(ModalRoutes.Explainer(explainerType, TripRouteAccents(route))) }
            } else {
                null
            }

        val onFollowTrip: (() -> Unit) = {
            openSheetRoute(
                SheetRoutes.TripDetails(
                    TripDetailsPageFilter(
                        tripId,
                        tripData.tripFilter.vehicleId,
                        trip.routeId,
                        trip.directionId,
                        tripFilter.stopId,
                        tripData.tripFilter.stopSequence,
                    )
                )
            )
        }

        TripDetails(
            trip,
            headerSpec,
            tripData.vehicleOnOtherTrip,
            onHeaderTap,
            ::onTapStop,
            onFollowTrip,
            onOpenAlertDetails,
            route,
            tripFilter,
            stopList,
            now,
            alertSummaries,
            globalResponse,
            isTripDetailsPage,
            modifier,
        )
    } else if (tripData != null && trip == null && tripData.vehicle != null) {
        MissingTripCard(
            MissingTripCardType.NotAvailable,
            routeAccents,
            Modifier.padding(horizontal = 16.dp),
        )
    } else {
        val placeholderTripInfo = LoadingPlaceholders.tripDetailsInfo()
        val placeholderTripStops = placeholderTripInfo.stops
        val placeholderTripId = placeholderTripInfo.trip.id

        val placeholderHeaderSpec =
            TripHeaderSpec.getSpec(
                placeholderTripId,
                placeholderTripInfo.stops,
                null,
                placeholderTripInfo.vehicle,
                placeholderTripInfo.vehicleStop,
            )

        CompositionLocalProvider(IsLoadingSheetContents provides true) {
            Column(modifier = modifier.loadingShimmer()) {
                TripDetails(
                    placeholderTripInfo.trip,
                    placeholderHeaderSpec,
                    false,
                    null,
                    onTapStop = {},
                    onFollowTrip = {},
                    onOpenAlertDetails = {},
                    placeholderTripInfo.route,
                    TripDetailsPageFilter("", "", Route.Id(""), 0, "", null),
                    placeholderTripStops,
                    now,
                    emptyMap(),
                    globalResponse
                        ?: GlobalResponse(ObjectCollectionBuilder("TripDetailsView.loading")),
                    isTripDetailsPage,
                )
            }
        }
    }
}

@Composable
fun TripDetails(
    trip: Trip,
    headerSpec: TripHeaderSpec?,
    vehicleOnOtherTrip: Boolean,
    onHeaderTap: (() -> Unit)?,
    onTapStop: (TripDetailsStopList.Entry) -> Unit,
    onFollowTrip: (() -> Unit),
    onOpenAlertDetails: (Alert) -> Unit,
    route: Route,
    tripFilter: TripDetailsPageFilter,
    stopList: TripDetailsStopList,
    now: EasternTimeInstant,
    alertSummaries: Map<String, AlertSummary?>,
    globalResponse: GlobalResponse,
    isTripDetailsPage: Boolean,
    modifier: Modifier = Modifier,
) {
    val routeAccents = TripRouteAccents(route)

    Column(modifier) {
        // If the loaded trip's ID does not match the trip ID of the vehicle,
        // it means that the selected trip is complete
        if (isTripDetailsPage && vehicleOnOtherTrip) {
            MissingTripCard(MissingTripCardType.Complete, routeAccents)
        } else {
            Column(Modifier.zIndex(1F)) {
                TripHeaderCard(
                    trip,
                    headerSpec,
                    tripFilter.stopId,
                    route,
                    routeAccents,
                    now,
                    onTap = onHeaderTap,
                    onFollowTrip = if (!isTripDetailsPage) onFollowTrip else null,
                )
            }
            Column(Modifier.offset(y = (-16).dp)) {
                TripStops(
                    tripFilter.stopId,
                    stopList,
                    tripFilter.stopSequence,
                    headerSpec,
                    now,
                    alertSummaries,
                    globalResponse,
                    onTapStop,
                    onOpenAlertDetails,
                    route,
                    routeAccents,
                )
            }
        }
    }
}
